tests/unit/s2n_self_talk_io_mem_test.c

```diff
@@ -219,6 +220,68 @@ int main(int argc, char **argv)
         EXPECT_SUCCESS(s2n_connection_free(server_conn));
     }
 
+    /* Test that dynamic buffers work correctly */
+    {
+        struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT);
+        EXPECT_NOT_NULL(client_conn);
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+        struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER);
+        EXPECT_NOT_NULL(server_conn);
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+        /* Enable the dynamic buffers setting */
+        EXPECT_SUCCESS(s2n_connection_set_dynamic_buffers(client_conn, true));
+        EXPECT_SUCCESS(s2n_connection_set_dynamic_buffers(server_conn, true));
+
+        /* Create nonblocking pipes */
+        struct s2n_test_io_pair io_pair;
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(client_conn, &io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(server_conn, &io_pair));
+
+        /* Do handshake */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+
+        s2n_blocked_status blocked = 0;
+        uint8_t buf[4096] = { 42 };
+
+        size_t sent_amount = 0;
+
+        while (sent_amount < s2n_array_len(buf)) {
+            int send_status = s2n_send(server_conn, &buf, s2n_array_len(buf) - sent_amount, &blocked);
+            EXPECT_SUCCESS(send_status);
+            sent_amount += s2n_array_len(buf);
+        }
```

Comment:
s2n_send actually returns the number of bytes it sent, so technically you shouldn't assume it sent all the data. Same with s2n_recv.

Should you also check that `conn->out` isn't freed on a partial write? You'll need to cause the write to block though, which means a custom io implementation or using stuffers for io.

```suggestion
            EXPECT_SUCCESS(send_status);
            sent_amount += send_status;
        }
```

Reply:
Ah i forgot to change that from what I had before.

I wasn't sure how to force `send` to reliably block. Do we have any other examples that you know of?

Reply:
The easiest way is to use the [testlib methods that set io to use stuffers](https://github.com/aws/s2n-tls/blob/main/tests/testlib/s2n_testlib.h#L42-L44). [Here](https://github.com/aws/s2n-tls/blob/main/tests/testlib/s2n_connection_test_utils.c#L65-L85)'s how that send is implemented.

Here's a [stuffer example](https://github.com/aws/s2n-tls/blob/main/tests/unit/s2n_early_data_io_test.c#L630) I wrote recently. Basically, if you don't provide enough space in the stuffer to write all the data, the write will fail.